pkg/math/oracle/median.go

```diff
@@ -0,0 +1,252 @@
+					"failed to scale price",
+					zap.Error(err),
+					zap.String("currency_pair", cp.String()),
+					zap.Int("decimals", cp.Decimals()),
+					zap.String("price", price.String()),
+				)
+
+				continue
+			}
+
+			scaledMedians[cp] = scaledPrice
+		}
+
+		// Determine the final aggregated price for each currency pair.
+		aggregatedMedians := make(map[oracletypes.CurrencyPair]*big.Int)
+		for _, cfg := range m.cfg.AggregatedFeeds {
+			// Get the converted prices for set of convertable markets.
+			// ex. BTC/USDT * USDT/USD = BTC/USD
+			//     BTC/USDC * USDC/USD = BTC/USD
+			convertedPrices := m.CalculateConvertedPrices(cfg, scaledMedians)
+
+			// If there were no converted prices, log an error and continue.
+			cp := cfg.CurrencyPair
+			if len(convertedPrices) == 0 {
+				m.logger.Debug("no converted prices", zap.String("currency_pair", cp.String()))
+				continue
+			}
+
+			// Take the median of the converted prices.
+			aggregatedMedians[cp] = aggregator.CalculateMedian(convertedPrices)
+			m.logger.Debug(
+				"calculated median price",
+				zap.String("currency_pair", cp.String()),
+				zap.String("price", aggregatedMedians[cp].String()),
+				zap.Any("converted_prices", convertedPrices),
+			)
+		}
+
+		// Scale all of the aggregated medians back to the original number of decimals.
+		for cp, price := range aggregatedMedians {
+			unscaledPrice, err := ScaleDownCurrencyPairPrice(int64(cp.Decimals()), price)
+			if err != nil {
+				m.logger.Error(
+					"failed to scale price",
+					zap.Error(err),
+					zap.String("currency_pair", cp.String()),
+					zap.Int("decimals", cp.Decimals()),
+					zap.String("price", price.String()),
+				)
+
+				continue
+			}
+
+			aggregatedMedians[cp] = unscaledPrice
+		}
+
+		return aggregatedMedians
+	}
+}
+
+// CalculateConvertedPrices calculates the converted prices for each currency pair using the
+// provided median prices and the conversion markets.
+//
+// For example, if the oracle receives a price for BTC/USDT and USDT/USD, it can use the conversion
+// market to convert the BTC/USDT price to BTC/USD. In this case, the medians map would contain
+// the median prices for BTC/USDT and USDT/USD, and the conversions would contain a sorted list of
+// operations to convert the price of BTC/USDT to BTC/USD i.e. BTC/USDT * USDT/USD = BTC/USD.
+func (m *MedianAggregator) CalculateConvertedPrices(
+	cfg config.AggregateFeedConfig,
+	medians map[oracletypes.CurrencyPair]*big.Int,
+) []*big.Int {
+	convertedPrices := make([]*big.Int, 0)
+	cp := cfg.CurrencyPair
+
+	for _, conversion := range cfg.Conversions {
+		// Ensure that the conversion is valid.
+		if err := config.CheckSort(cp, conversion); err != nil {
+			m.logger.Error(
+				"invalid conversion",
+				zap.Error(err),
+				zap.Any("conversions", conversion),
+			)
+
+			continue
+		}
+
+		// Calculate the converted price.
+		convertedPrice, err := m.CalculateConvertedPrice(cp, conversion, medians)
+		if err != nil {
```

Comment:
we should def add test cases for this codepath

Reply:
the `MedianAggregator` should never be constructed if a bad path is included. Ill some for posterity.